pkg/apis/auditregistration/validation/validation.go

```diff
@@ -0,0 +1,138 @@
+	var allErrs field.ErrorList
+	if b.ThrottleConfig != nil {
+		allErrs = append(allErrs, ValidateWebhookThrottleConfig(b.ThrottleConfig, fldPath.Child("throttleConfig"))...)
+	}
+	allErrs = append(allErrs, ValidateWebhookClientConfig(&b.ClientConfig, fldPath.Child("clientConfig"))...)
+	return allErrs
+}
+
+// ValidateWebhookThrottleConfig validates the throttle config
+func ValidateWebhookThrottleConfig(c *auditregistration.WebhookThrottleConfig, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	if c.ThrottleQPS != nil && *c.ThrottleQPS <= 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("throttleQPS"), c.ThrottleQPS, "throttleQPS must be a positive number"))
+	}
+	if c.ThrottleBurst != nil && *c.ThrottleBurst <= 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("throttleBurst"), c.ThrottleBurst, "throttleBurst must be a positive number"))
+	}
+	return allErrs
+}
+
+// ValidateWebhookClientConfig validates the WebhookClientConfig
+func ValidateWebhookClientConfig(c *auditregistration.WebhookClientConfig, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+
+	if c.URL == nil && c.Service == nil {
+		allErrs = append(allErrs, field.Invalid(fldPath, c, "either url or service definition must be set for webhook"))
+	}
+	if c.URL != nil && c.Service != nil {
+		allErrs = append(allErrs, field.Invalid(fldPath, c, "both url and service definition should not be set, choose one"))
+	}
+	return allErrs
+}
+
+// ValidatePolicy validates the audit policy
+func ValidatePolicy(policy *auditregistration.Policy) field.ErrorList {
+	var allErrs field.ErrorList
+	allErrs = append(allErrs, validateOmitStages(policy.OmitStages, field.NewPath("omitStages"))...)
+	if policy.Level != nil {
+		allErrs = append(allErrs, validateLevel(*policy.Level, field.NewPath("level"))...)
+	}
+	return allErrs
+}
+
+var validLevels = []string{
+	string(auditregistration.LevelNone),
+	string(auditregistration.LevelMetadata),
+	string(auditregistration.LevelRequest),
+	string(auditregistration.LevelRequestResponse),
+}
+
+var validOmitStages = []string{
+	string(auditregistration.StageRequestReceived),
+	string(auditregistration.StageResponseStarted),
+	string(auditregistration.StageResponseComplete),
+	string(auditregistration.StagePanic),
+}
+
+func validateLevel(level auditregistration.Level, fldPath *field.Path) field.ErrorList {
+	switch level {
+	case auditregistration.LevelNone, auditregistration.LevelMetadata, auditregistration.LevelRequest, auditregistration.LevelRequestResponse:
+		return nil
+	case "":
+		return field.ErrorList{field.Required(fldPath, "")}
+	default:
+		return field.ErrorList{field.NotSupported(fldPath, level, validLevels)}
+	}
+}
+
+func validateOmitStages(omitStages []auditregistration.Stage, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	for i, stage := range omitStages {
```

Comment:
Is it OK to omit all stages?

Reply:
omit all stages == audit level none. As the intention is for this to eventually define the default behavior (when no other policy rules are matched), it makes sense to allow specifying none here. Also, if you set the audit level to none, it would be weird to still require users to set some stage. I suppose we could require it to be non-empty IFF level != none.

Reply:
Personally I'd lean towards allowing them to say whatever they want, and explaining the (lack of) consequences in the status or via events, but maybe that's more complex than you want.

It is currently way easier to relax validation than to tighten it, so you may want to err on the side of being a little pedantic.